cmake/KokkosConfig.cmake.in

```diff
@@ -62,3 +62,7 @@ ELSEIF(@Kokkos_ENABLE_CUDA@ AND NOT "separable_compilation" IN_LIST Kokkos_FIND_
 ENDIF()
 
 set(Kokkos_COMPILE_LANGUAGE @KOKKOS_COMPILE_LANGUAGE@)
+
+# Kokkos_<INFO> vars
+
+@KOKKOS_OPTION_SET_STATEMENTS_STR@
```

Comment:
What variables actually need to be set and we are not already exporting?

Reply:
> I can see a desire to down-select the set of vars that Kokkos exports. But note that in the absences of these variables, downstream customers are just grepping the installed [KokkosCore_config.h](https://github.com/bartlettroscoe/kokkos/blob/tril-11545-kokkos-use-native-cmake-develop/cmake/KokkosCore_config.h.in) file (for example, as you can see see [here](https://github.com/sandialabs/Albany/blob/3dfea62da5000fef36cfacf13a2871dfb850b9f8/CMakeLists.txt#L217-L230)).

We export `Kokkos_ENABLE_{OPENMP,CUDA,...}` when the backends are enabled.  This is unnecessary.

Reply:
What do you mean?

Reply:
That users can do
```CMake
find_package(Kokkos REQUIRED)
if(Kokkos_ENABLE_CUDA)
  // ...
endif()

Reply:
See https://github.com/kokkos/kokkos/blob/4197fa833d27679fecb8a22a3321582340417181/cmake/KokkosConfigCommon.cmake.in#L1-L7.

Reply:
Are those same CMake vars provided for CMake customers that pull in Kokkos with `add_subdirectory(kokkos)`?  The way you use a package needs to be identical regardless if it is pulled in with `find_package(<Package>)` or `add_subdirectory(<package>)`.

It may be that downstream TriBITS packages like Tpetra need to be refactored to use a set of variables that Kokkos means to export and not just what cache vars it provides and are accessible with `add_subdirectory(kokkos)`.  Current, in Trilinos, every `Kokkos_<INFO>` cache var is accessible because they are all globally accessible.

Reply:
> That users can do
> 
> ```cmake
> find_package(Kokkos REQUIRED)
> if(Kokkos_ENABLE_CUDA)
>   // ...
> endif()
> ```

You are saying that many of these vars are already exported by `KokkosConfig.cmake` before this PR?

Reply:
Yes.  From my quickly skimming through Trilinos, we are only missing `Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE`.

Reply:
From my memory, there are more variables than that.  For example, where is `Kokkos_ENABLE_CUDA` set in `KokkosConfig.cmake` or `KokkoConfigCommon.cmake`?

Note that some Trilinos packages use `assert_defined()` to ensure that a variable so not setting a var if it is not true is not an option.

Reply:
> From my memory, there are more variables than that. For example, where is Kokkos_ENABLE_CUDA set in KokkosConfig.cmake or KokkoConfigCommon.cmake?

We set all the `Kokos_ENABLE_<BACKEND>` variable here https://github.com/kokkos/kokkos/blob/3329989db9fddbdc90bbe204df1bd17ac22c5594/cmake/KokkosConfigCommon.cmake.in#L10-L12

> Note that some Trilinos packages use [assert_defined()](https://tribitspub.github.io/TriBITS/users_guide/index.html#assert-defined) to ensure that a variable is defined before checking it so not setting a var if it is not true is not an option. That avoid accidentally misspelling a var name.

We asked them to drop these (See trilinos/Trilinos#11896)

Reply:
> Yes. From my quickly skimming through Trilinos, we are only missing `Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE` that is used in two places.
> 
> ```
> packages/phalanx/CMakeLists.txt:122:    IF(NOT Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE)
> packages/stokhos/src/CMakeLists.txt:1098:      IF (Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE)
> ```

@dalg24 and @masterleinad,

So where and how should `Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE` be set in `KokkosConfig[Common].cmake`?  

After the merge of:

* https://github.com/trilinos/Trilinos/pull/11896

I can merge Trilinos 'develop' into:

* https://github.com/trilinos/Trilinos/pull/11863

and try removing the commits:

* https://github.com/trilinos/Trilinos/pull/11863/commits/ecc56096e4744c450f94bc333583272e93ee82d2
* https://github.com/kokkos/kokkos/pull/6138/commits/54425648d2974a99a78e3825fe3d7465dc0a388d

and related commits.

But if `KokkosConfig.cmake` is not defining `Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE`  then it will always be implicitly `OFF` in CUDA configurations when building downstream packages.

Reply:
> But if `KokkosConfig.cmake` is not defining `Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE` then it will always be implicitly `OFF` in CUDA configurations when building downstream packages.

Also, if you look at Albany [here](https://github.com/sandialabs/Albany/blob/3dfea62da5000fef36cfacf13a2871dfb850b9f8/CMakeLists.txt#L217-L230), you see the following variables should be set as well:

* `KOKKOS_COMPILER_CUDA_VERSION`

Users should not have to grep the installed `KokkosCore_config.h` file to get info at configure time.

Reply:
> So where and how should Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE be set in KokkosConfig[Common].cmake?

We will handle shortly.  Will keep you updated.

Reply:
@dalg24, I did a more detailed search for `Kokkos_<???>` vars used in Trilinos (on the branch [11545-kokkos-use-native-cmake](https://github.com/trilinos/Trilinos/tree/11545-kokkos-use-native-cmake) in PR https://github.com/trilinos/Trilinos/pull/11863 as of commit https://github.com/trilinos/Trilinos/pull/11863/commits/a38a94477a838461262f5ddd47ab9ebc6504c70a) using:

```
$ cd Trilinos/

$  find packages/ \( -name CMakeLists.txt -or -name "*.cmake" \) -exec grep -nH Kokkos_ {} \; \
   &> ../kokkos_vars_used_by_trilinos.txt

$ cat junk/kokkos_vars_used_by_trilinos.txt \
  | grep -v "^packages/compadre/" |  grep -v "Kokkos_ENABLE_[$]" \
  | grep -v "[.]hpp" | grep -v "[.]h" | grep -v "[.]cpp" \
  | sed 's/.*[^a-zA-Z0-9_]\(Kokkos_[a-zA-Z0-9_][a-zA-Z0-9_]*\).*/\1/' | grep -v "^packages" \
  | sort | uniq
```

and it gave the list:

```
Kokkos_ARCH
Kokkos_ARCH_INTEL_PVC
Kokkos_BlockCGEx
Kokkos_BlockGmresEx
Kokkos_DIR
Kokkos_ENABLE_CUDA
Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE
Kokkos_ENABLE_CUDA_UVM
Kokkos_ENABLE_DEBUG
Kokkos_ENABLE_HIP
Kokkos_ENABLE_OpenMP
Kokkos_ENABLE_OPENMP
Kokkos_ENABLE_OpenMPTarget
Kokkos_ENABLE_OPENMPTARGET
Kokkos_ENABLE_PTHREAD
Kokkos_ENABLE_SERIAL
Kokkos_ENABLE_SYCL
Kokkos_ENABLE_THREADS
Kokkos_INCLUDE_DIRS
Kokkos_VERSION
```

The full grep before filtering is given in [kokkos_vars_used_by_trilinos.txt](https://github.com/kokkos/kokkos/files/11502583/kokkos_vars_used_by_trilinos.txt).

Some of those are not Kokkos variables (like `Kokkos_BlockGmresEx` and `Kokkos_BlockCGEx`) but many are.  And some like `Kokkos_INCLUDE_DIRS` need to be removed from Trilinos.

NOTE: I removed the matches under `packages/compadre/` because `packages/compadre/CMakeLists.txt` defines a bunch of options that start with `Kokkos_` so that gives false matches for true Kokkos configure vars.

P.S. I am guessing that some of these with upper-case options like `Kokkos_ENABLE_OPENMP` need to be fixed?